File: Source/Samples/95_Periodic/code_apulldownmenu.cpp
// APulldownMenu application source code
#include <Urho3D/AUI/APulldownMenu.h>
#include "PeriodicApp.h"

void PeriodicApp::setup_apulldownmenu( AWidget *layout )
{
    unsigned ii = 0;
    PODVector<AWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for ( ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleApulldownmenuEvent ));
    layout->SearchWidgetClass( "TBPulldownMenu", dest );
    for ( ii = 0;  ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleApulldownmenuEvent ));
}

void PeriodicApp::HandleApulldownmenuEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;
    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uipulldownmenucode" )
        {
            AppLog( "APulldownMenu support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Sources/code_apulldownmenu.cpp", widget->GetParent() );
        }
        if (widget->GetId() ==  "uipulldownmenulayout" )
        {
            AppLog( "APulldownMenu support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_apulldownmenu.ui.txt", widget->GetParent() );
        }
    }
    else if (eventData[P_TYPE] == UI_EVENT_TYPE_CHANGED )
    {
        if (widget->GetId() ==  "FileMenu" )
        {
            APulldownMenu *pdm = static_cast<APulldownMenu*>(widget);
            if ( pdm)
                AppLog( "APulldownMenu event : " + widget->GetId() + " selected entry = " + String( pdm->GetSelectedId()) );
        }
        if (widget->GetId() ==  "EditMenu" )
        {
            APulldownMenu *pdm = static_cast<APulldownMenu*>(widget);
            if ( pdm)
                AppLog( "APulldownMenu event : " + widget->GetId() + " selected entry = " + String( pdm->GetSelectedId()) );
        }
        if (widget->GetId() ==  "ViewMenu" )
        {
            APulldownMenu *pdm = static_cast<APulldownMenu*>(widget);
            if ( pdm)
                AppLog( "APulldownMenu event : " + widget->GetId() + " selected entry = " + String( pdm->GetSelectedId()) );
        }
        if (widget->GetId() ==  "HelpMenu" )
        {
            APulldownMenu *pdm = static_cast<APulldownMenu*>(widget);
            if ( pdm)
                AppLog( "APulldownMenu event : " + widget->GetId() + " selected entry = " + String( pdm->GetSelectedId()) );
        }
    }
}

File: bin/Data/AUI/Periodic/LuaScripts/code_apulldownmenu.lua
-- APulldownMenu application source code


function setup_uipulldownmenu( layout )
    local dest = layout:SearchAWidgetClass( "TBButton" )
    for i, awid in ipairs(dest) do
        SubscribeToEvent(awid, "WidgetEvent", "HandleUipulldownmenuEvent")
    end

    local dest1 = layout:SearchAWidgetClass( "TBPulldownMenu" )
    for i, apul in ipairs(dest1) do
        SubscribeToEvent(apul, "WidgetEvent", "HandleUipulldownmenuEvent")
    end
end

function HandleUipulldownmenuEvent( eventType, eventData)
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
        return
    end
    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if widget:GetId() == "uipulldownmenucode" then
            AppLog( "APulldownMenu support : " .. widget:GetId() .. " was pressed " )
            ViewCode ( "LuaScripts/code_apulldownmenu.lua", widget:GetParent() )
        end
        if widget:GetId() ==  "uipulldownmenulayout" then
            AppLog( "APulldownMenu support : " .. widget:GetId() .. " was pressed ")
            ViewCode ( "Scenes/layout_apulldownmenu.ui.txt", widget:GetParent() )
        end
    elseif eventData["Type"]:GetInt() == UI_EVENT_TYPE_CHANGED then
        if widget:GetId() == "FileMenu" then
            local pdm = tolua.cast(widget, "APulldownMenu") -- HELP No conversion from 'AWidget' to 'APulldownMenu'
            if pdm ~= nil then
                AppLog( "APulldownMenu event : " .. widget:GetId() .. " selected entry = " .. pdm:GetSelectedId())
            end
        end
        if widget:GetId() == "EditMenu" then
            local pdm = tolua.cast(widget, "APulldownMenu")
            if pdm ~= nil then
                AppLog( "APulldownMenu event : " .. widget:GetId() .. " selected entry = " .. pdm:GetSelectedId())
            end
        end
        if widget:GetId() == "ViewMenu" then
            local pdm = tolua.cast(widget, "APulldownMenu")
            if pdm ~= nil then
                AppLog( "APulldownMenu event : " .. widget:GetId() .. " selected entry = " ..  pdm:GetSelectedId())
            end
        end
        if widget:GetId() == "HelpMenu" then
            local pdm = tolua.cast(widget, "APulldownMenu")
            if pdm ~= nil then
               AppLog( "APulldownMenu event : " .. widget:GetId() .. " selected entry = " .. pdm:GetSelectedId())
            end
        end
    end
end

